cpu/nrf52: i2c: add support for 16-bit register addresses

nicoHarel opened the pull request:
### Contribution description

Hello everyone.

This PR is related to issue: https://github.com/RIOT-OS/RIOT/issues/16680

It adds i2C 16-bits register addressing support to the NRF52 CPUs. I guess this mainly add compatibility for EEPROM ICs. 

### Testing procedure

Tested with a NRF52840, a AT24C256C EEPROM, and the according RIOT test script: tests/driver_at24cxxx.

With the "erase flag" set I've got the following output: 
> 2021-08-05 11:00:35,074 # main(): This is RIOT! (Version: 2021.10-devel-293-g1b015-cpu/nrf52/periph/i2c)
2021-08-05 11:00:35,075 # Starting tests for module at24cxxx
2021-08-05 11:00:35,077 # EEPROM size: 32768 byte
2021-08-05 11:00:35,078 # Page size  : 64 byte
2021-08-05 11:00:35,079 # [SUCCESS] at24cxxx_init
2021-08-05 11:00:41,159 # [SUCCESS] at24cxxx_erase
2021-08-05 11:00:41,159 # [SUCCESS] at24cxxx_write_byte
2021-08-05 11:00:41,160 # [SUCCESS] at24cxxx_read_byte
2021-08-05 11:00:41,160 # [SUCCESS] write_byte/read_byte
2021-08-05 11:00:41,161 # [SUCCESS] at24cxxx_write
2021-08-05 11:00:41,161 # [SUCCESS] at24cxxx_read
2021-08-05 11:00:41,161 # [SUCCESS] write/read
2021-08-05 11:00:41,161 # [SUCCESS] at24cxxx_set
2021-08-05 11:00:41,162 # [SUCCESS] set/read
2021-08-05 11:00:41,162 # Finished tests for module at24cxxx

Nothing should have changed for 8 bits devices. My DS3231 is still working. 

### Issues/PRs references

Fixes #https://github.com/RIOT-OS/RIOT/issues/16680

Have a good day =)

benpicco commented on cpu/nrf52/periph/i2c.c:
```diff
@@ -169,27 +169,44 @@ int i2c_write_regs(i2c_t dev, uint16_t addr, uint16_t reg,
      *
      * CAUTION: this might become critical when transferring large blocks of
      *          data as the temporary buffer is allocated on the stack... */
-    uint8_t buf_tmp[len + 1];
```
urgh so this was incidentally putting up to 256 bytes of memory on the stack and using VLA on top?
I know this was not introduced by you, but to avoid nasty surprises I would make this a 

```C
    static uint8_t tx_buf[256];
```

nicoHarel replied:
Yeah the VLA it is not the best solution in my opinion either... 

The fixed size buffer at the top seems more reasonable. There's already a size check on the data length parameter so I don't see any reason why we couldn't do this instead.

benpicco commented on cpu/nrf52/periph/i2c.c:
```diff
@@ -169,27 +169,44 @@ int i2c_write_regs(i2c_t dev, uint16_t addr, uint16_t reg,
+        reg_addr_len = 2;
+        /* Prepare the 16-bit register transfer  */
+        uint8_t buf_tmp[reg_addr_len + len];
+        buf_tmp[0] = reg>>8; /* AddrH in the first byte  */
+        buf_tmp[1] = reg&0xFF; /* AddrL in the second byte  */
```
```suggestion
        buf_tmp[0] = reg >> 8; /* AddrH in the first byte  */
        buf_tmp[1] = reg & 0xFF; /* AddrL in the second byte  */
```

benpicco commented on cpu/nrf52/periph/i2c.c:
```diff
@@ -169,27 +169,44 @@ int i2c_write_regs(i2c_t dev, uint16_t addr, uint16_t reg,
+        reg_addr_len = 1;
+        uint8_t buf_tmp[reg_addr_len + len];
+        buf_tmp[0] = reg;
+        memcpy(&buf_tmp[reg_addr_len], data, len);
+        return i2c_write_bytes(dev, addr, buf_tmp, (reg_addr_len + len), flags);
```
looks like you can pull this out of the if block

benpicco commented on cpu/nrf52/periph/i2c.c:
```diff
@@ -212,21 +229,36 @@ int i2c_read_regs(i2c_t dev, uint16_t addr, uint16_t reg,
-    bus(dev)->TXD.PTR = (uint32_t)&reg;
-    bus(dev)->TXD.MAXCNT = 1;
+    if (flags & (I2C_REG16)) {
+        buf_regs[0] = reg>>8; /* AddrH in the first byte  */
+        buf_regs[1] = reg&0xFF; /* AddrL in the second byte  */
+        bus(dev)->TXD.PTR = (uint32_t)buf_regs;
+        bus(dev)->TXD.MAXCNT = 2;
+    }
+    else {
+        bus(dev)->TXD.PTR = (uint32_t)&reg;
+        bus(dev)->TXD.MAXCNT = 1;
+    }
```
Won't this suffice?

```suggestion
        reg = htons(reg);
        bus(dev)->TXD.MAXCNT = 2;
    }
    else {
        bus(dev)->TXD.MAXCNT = 1;
    }
    bus(dev)->TXD.PTR = (uint32_t)&reg;
```

nicoHarel replied:
Just tried it, it seems fine yes.

benpicco reviewed (commented):
Thank you for the patch!
Looks good. The use of VLAs is unfortunate but it's already present in the existing driver.

Maybe we can get rid of that.

nicoHarel commented:
Updated. Thanks for the review :+1: 

I removed the VLA and now there's a fixed size buffer at the top. But it means that, as soon as we use the i2C on NRF52, it's going to take up some space. It's less "dangerous" but, you're sure this is better ? I don't care either way, just asking out of curiosity.

benpicco commented on cpu/nrf52/periph/i2c.c:
```diff
@@ -167,38 +173,31 @@ void i2c_release(i2c_t dev)
-     * CAUTION: this might become critical when transferring large blocks of
-     *          data as the temporary buffer is allocated on the stack... */
+    /* Resets the TX buffer */
+    memset(tx_buf, 0, sizeof(tx_buf));
```
That should not be needed, we already overwrite it with `memcpy` and do not transfer more bytes than written to it.

benpicco commented on cpu/nrf52/periph/i2c.c:
```diff
@@ -212,21 +228,35 @@ int i2c_read_regs(i2c_t dev, uint16_t addr, uint16_t reg,
            (int)len, (int)reg, (int)addr);
 
+    /* Buffer used to split the register address for 16-bit addressing */
+    uint8_t buf_regs[2];
```
this is now unused

benpicco commented:
Yea this is unfortunate. We could make the buffer size configurable, but even with the VLA solution the memory must come from somewhere (ok, unless we never do 255 byte reads).

The [SDK driver](https://github.com/NordicSemiconductor/nrfx/blob/master/drivers/src/nrfx_twi.c) seems to be doing fine without an additional buffer, so might well be that we are doing something wrong here.

nicoHarel commented:
Okay, removed the unnecessary lines. 

I tried to have a quick look to see if we can do something about that buffer, but I land on the same problem people before me had . I can't seem to find a way to transfer the rest of the data (after the register address) without an other start condition (thus breaking the i2C protocol). I can't find a shortcut in the nrf52840 bitfields to disable the start condition... 

Anyway, if it's all good for this PR, I'll squash.

benpicco reviewed (commented):
Please squash!

And while you're at it, please change the commit message to something like

    cpu/nrf52: i2c: add support for 16-bit register addresses

benpicco commented on cpu/nrf52/periph/i2c.c:
```diff
@@ -167,29 +173,36 @@ void i2c_release(i2c_t dev)
+    }
+
+    memcpy(&tx_buf[reg_addr_len], data, len);
+    return i2c_write_bytes(dev, addr, tx_buf, (reg_addr_len + len), flags);
```
```suggestion
    return i2c_write_bytes(dev, addr, tx_buf, reg_addr_len + len, flags);
```

benpicco commented on cpu/nrf52/periph/i2c.c:
```diff
@@ -167,29 +173,36 @@ void i2c_release(i2c_t dev)
+    if (flags & (I2C_NOSTART | I2C_ADDR10)) {
         return -EOPNOTSUPP;
     }
-    /* the nrf52's TWI device does not support to do two consecutive transfers
-     * without a repeated start condition in between. So we have to put all data
-     * to be transferred into a temporary buffer
-     *
-     * CAUTION: this might become critical when transferring large blocks of
-     *          data as the temporary buffer is allocated on the stack... */
```
We should retain this comment so people know why this buffer is there in the first place.

nicoHarel replied:
I think it should be modified then.

There's no temporary buffer allocated, the buffer is permanent and has a fixed size whether or not we transfer large blocks.

benpicco commented on cpu/nrf52/periph/i2c.c:
```diff
@@ -40,6 +41,11 @@
+/**
+ * @brief   Allocate for a tx buffer
+ */
+static uint8_t tx_buf[256];
```
We should probably protect access to this with a mutex

nicoHarel commented:
Okay, I've put the buffer comment back (with a few mods). And also added a mutex for the tx buffer. But do we really needed one ? I mean, isn't the user supposed to use the i2c_acquire and i2c_release functions ?

benpicco commented:
> But do we really need one ? I mean, isn't the user supposed to use the i2c_acquire and i2c_release functions?

Yes but the buffer is shared between all I2C busses

benpicco commented on cpu/nrf52/periph/i2c.c:
```diff
@@ -167,29 +178,51 @@ void i2c_release(i2c_t dev)
-    buf_tmp[0] = reg;
-    memcpy(&buf_tmp[1], data, len);
-    return i2c_write_bytes(dev, addr, buf_tmp, (len + 1), flags);
+     * CAUTION: this might be critical as the buffer is allocated on the stack.
+     *  */
```
But now the buffer is no longer allocated on the stack :wink:

nicoHarel replied:
Damn it, sorry, I'll fix that

benpicco commented on cpu/nrf52/periph/i2c.c:
```diff
@@ -167,29 +178,51 @@ void i2c_release(i2c_t dev)
-    buf_tmp[0] = reg;
-    memcpy(&buf_tmp[1], data, len);
-    return i2c_write_bytes(dev, addr, buf_tmp, (len + 1), flags);
+     * CAUTION: this might be critical as the buffer is allocated on the stack.
+     *  */
```
But now the buffer is no longer allocated on the stack :wink:

benpicco reviewed (approved):
Looks good, thank you for fixing this!
